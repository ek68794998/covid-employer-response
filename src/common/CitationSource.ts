export class CitationSource {
	public date?: string;

	public link: string = "";

	public source: string = "";

	public title: string = "";
}
