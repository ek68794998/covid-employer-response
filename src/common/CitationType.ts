export const CitationTypeValues: string[] = [
	"hearsay",
	"statement",
	"publication",
];

export type CitationType = typeof CitationTypeValues[number];

export const getCitationTypeValue = (type?: CitationType): number =>
	type ? CitationTypeValues.indexOf(type) : -1;
