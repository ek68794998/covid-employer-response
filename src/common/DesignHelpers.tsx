import React from "react";

export class DesignHelpers {
	public static materialIcon(iconName: string, className?: string): JSX.Element {
		return <i className={`material-icons ${className || ""}`}>{iconName}</i>;
	}
}
