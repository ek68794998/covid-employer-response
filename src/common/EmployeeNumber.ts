type EmployeeMagnitude = 6 | 9 | null;

export class EmployeeNumber {
	public exactValue: number;

	public magnitude: EmployeeMagnitude;

	public value: number;

	public constructor(value: number) {
		this.exactValue = value;

		this.magnitude = EmployeeNumber.getEmployeeMagnitude(value);

		this.value =
			this.magnitude
				? Math.round(1234567 / Math.pow(10, this.magnitude - 1)) / 10
				: value;
	}

	public get magnitudeLabel(): string | null {
		return this.magnitude === 6 ? "million" : this.magnitude === 9 ? "billion" : null;
	}

	public toString(): string {
		return this.magnitude ? `${this.value} ${this.getEmployeeMagnitudeName()}` : this.value.toLocaleString();
	}

	private static getEmployeeMagnitude(value: number): EmployeeMagnitude {
		if (!value) {
			return null;
		}

		if (value >= 1000000000) {
			return 6;
		}

		if (value >= 1000000) {
			return 9;
		}

		return null;
	}

	private getEmployeeMagnitudeName(): string | null {
		switch (this.magnitude) {
			case 6: return "million";
			case 9: return "billion";
			default: return null;
		}
	}
}