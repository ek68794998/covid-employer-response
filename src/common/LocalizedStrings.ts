export interface LocalizedStrings { [key: string]: LocalizedStringValue }

export const format = (translation: string, args: { [key: string]: any }): string => {
	if (!translation) {
		throw new Error("Translated string must not be empty.");
	}

	return translation.replace(/\{([^\}]+)\}/g, (value: string, argName: string) => args[argName] || argName);
};

// Would prefer to use 'string | { [key: string]: LocalizedStringValue }' here.
// However, then TypeScript complains if you do 'strings.foo.bar', saying that 'bar' does not exist on type 'string'.
type LocalizedStringValue = any;