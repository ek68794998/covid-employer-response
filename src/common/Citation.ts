import { CitationSource } from "./CitationSource";
import { CitationType } from "./CitationType";

export class Citation {
	public positivity: number = 0;

	public sources?: CitationSource[];

	public summary: string = "";

	public type?: CitationType;
}
