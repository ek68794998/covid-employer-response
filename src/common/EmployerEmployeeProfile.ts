import { EmployeeNumber } from "./EmployeeNumber";

export type EmployerEmployeeProfileType =
	"range"
	| "approximately"
	| "exactly";

export class EmployerEmployeeProfile {
	public lowerBound?: number;

	public type: EmployerEmployeeProfileType = "exactly";

	public upperBound: number = 0;

	public year: number = 0;

	public yearQuarter?: "Q1" | "Q2" | "Q3" | "Q4";

	public static toString(p: EmployerEmployeeProfile, includeDate: boolean, useShortText: boolean): string | null {
		const lowerBound: EmployeeNumber | null =
			(p.lowerBound && p.lowerBound > 0) ? new EmployeeNumber(p.lowerBound) : null;

		const upperBound: EmployeeNumber | null =
			p.upperBound > 0 ? new EmployeeNumber(p.upperBound) : null;

		let employeeCountString: string;

		switch (p.type) {
			case "approximately":
				if (!upperBound) {
					throw new Error("Invalid value for approximate string conversion.");
				}

				employeeCountString = upperBound.toString(useShortText);

				if (!upperBound.getMagnitudeLabel()) {
					employeeCountString = `~${employeeCountString}`;
				}

				break;

			case "range":
				if (lowerBound && upperBound) {
					if (lowerBound.exactValue > upperBound.exactValue) {
						throw new Error("Lower bound is greater than or equal to upper bound specified for range string conversion.");
					}

					employeeCountString =
						(lowerBound.magnitude && lowerBound.magnitude === upperBound.magnitude)
							? `${lowerBound.getAdjustedValue()} ${String.fromCharCode(0x2013)} ${upperBound.toString(useShortText)}`
							: `${lowerBound.toString(useShortText)} ${String.fromCharCode(0x2013)} ${upperBound.toString(useShortText)}`;
				} else if (lowerBound) {
					employeeCountString =
						useShortText
							? `${String.fromCharCode(0x2265)} ${lowerBound.toString(useShortText)}`
							: `More than ${lowerBound.toString(useShortText)}`;
				} else if (upperBound) {
					employeeCountString =
						useShortText
							? `${String.fromCharCode(0x2264)} ${upperBound.toString(useShortText)}`
							: `Less than ${upperBound.toString(useShortText)}`;
				} else {
					throw new Error("Invalid lower and upper bounds specified for range string.");
				}

				break;

			case "exactly":
			default:
				if (!upperBound) {
					throw new Error("Invalid value for default string conversion.");
				}

				employeeCountString = upperBound.toString(useShortText);

				break;
		}

		if (includeDate && p.year > 0) {
			const dateString: string =
				p.yearQuarter ? `${p.yearQuarter} ${p.year}` : `${p.year}`;

			employeeCountString = `${employeeCountString} (${dateString})`;
		}

		return employeeCountString;
	}
}
