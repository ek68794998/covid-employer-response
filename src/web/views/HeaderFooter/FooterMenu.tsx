import React from "react";
import { useSelector } from "react-redux";

import { AppState } from "../../state/AppState";
import { getStrings } from "../../state/ducks/localization/selectors";

import { AuthorUrl, ProjectUrl } from "../../../common/constants/UrlConstants";
import { LocalizedStrings } from "../../../common/LocalizedStrings";

import "./HeaderFooter.scss";

const HeaderMenu: React.FC = (): React.ReactElement => {
	const strings: LocalizedStrings = useSelector(getStrings);

	return (
		<footer>
			<div className="HeaderFooter__Container">
				<span id="maintained-by">
					Maintained by <a href={AuthorUrl}>Eric Kumlin</a>
				</span>
				<span id="project-source">
					<em>{strings.appTitle}</em> is open-source and open-data, hosted on <a href={ProjectUrl}>GitHub</a>.
				</span>
			</div>
		</footer>
	);
};

export default HeaderMenu as any;
