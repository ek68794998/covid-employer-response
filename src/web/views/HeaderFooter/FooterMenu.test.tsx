import React from "react";
import { render } from "react-dom";
import { BrowserRouter } from "react-router-dom";

import { LocalizedStrings } from "../../../common/LocalizedStrings";

import FooterMenu from "./FooterMenu";

jest.mock("react-redux", () => ({
	useSelector: (): {} => ({
		getStrings: (): LocalizedStrings => ({
			appTitle: "App Title",
		}),
	}),
}));

describe("<FooterMenu />", () => {
	test("renders without exploding", () => {
		const div: HTMLDivElement = document.createElement("div");

		render(
			<BrowserRouter>
				<FooterMenu.WrappedComponent />
			</BrowserRouter>,
			div,
		);
	});
});
