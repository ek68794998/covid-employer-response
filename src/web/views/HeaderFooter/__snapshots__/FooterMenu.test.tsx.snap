// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FooterMenu /> renders without exploding 1`] = `
<footer>
  <div
    className="HeaderFooter__Container"
  >
    <span
      id="maintained-by"
    >
      Maintained by 
      <a
        href="https://github.com/ekumlin"
      >
        Eric Kumlin
      </a>
    </span>
    <span
      id="project-source"
    >
      <em>
        App Title
      </em>
       is open-source and open-data, hosted on 
      <a
        href="https://github.com/ekumlin/covid-employer-response"
      >
        GitHub
      </a>
      .
    </span>
  </div>
</footer>
`;
